checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -0,0 +1,1606 @@
+   * type {@code ImmutableSet<LocalVarWithTree>}, each representing a set of resource aliases for
+   * some value with a non-empty {@code @MustCall} obligation.
+   *
+   * @param cfg the control flow graph of the method to check
+   */
+  // TODO: This analysis is currently implemented directly using a worklist; in the future, it
+  // should be transitioned to use the Checker dataflow framework.
+  /* package-private */
+  void analyze(ControlFlowGraph cfg) {
+    Set<BlockWithFacts> visited = new LinkedHashSet<>();
+    Deque<BlockWithFacts> worklist = new ArrayDeque<>();
+
+    // add any owning parameters to initial set of variables to track
+    BlockWithFacts entryBlockFacts =
+        new BlockWithFacts(cfg.getEntryBlock(), computeOwningParameters(cfg));
+    worklist.add(entryBlockFacts);
+    visited.add(entryBlockFacts);
+
+    while (!worklist.isEmpty()) {
+      BlockWithFacts curBlockFacts = worklist.removeLast();
+      List<Node> nodes = curBlockFacts.block.getNodes();
+      // A *mutable* set that eventually holds the set of facts to be propagated to successor
+      // blocks. The set is initialized to the current facts and updated by the methods invoked in
+      // the for loop below.
+      Set<ImmutableSet<LocalVarWithTree>> newFacts = new LinkedHashSet<>(curBlockFacts.facts);
+
+      for (Node node : nodes) {
+        if (node instanceof AssignmentNode) {
+          handleAssignment((AssignmentNode) node, newFacts);
+        } else if (node instanceof ReturnNode) {
+          handleReturn((ReturnNode) node, cfg, newFacts);
+        } else if (node instanceof MethodInvocationNode || node instanceof ObjectCreationNode) {
+          handleInvocation(newFacts, node);
+        }
+      }
+
+      handleSuccessorBlocks(visited, worklist, newFacts, curBlockFacts.block);
+    }
+  }
+
+  /**
+   * Update a set of facts to account for a method or constructor invocation
+   *
+   * @param facts the facts to update
+   * @param node the method or constructor invocation
+   */
+  private void handleInvocation(Set<ImmutableSet<LocalVarWithTree>> facts, Node node) {
+    doOwnershipTransferToParameters(facts, node);
+    // Count calls to @CreatesMustCallFor methods as creating new resources, for now.
+    if (node instanceof MethodInvocationNode
+        && typeFactory.canCreateObligations()
+        && typeFactory.hasCreatesMustCallFor((MethodInvocationNode) node)) {
+      checkCreatesMustCallForInvocation(facts, (MethodInvocationNode) node);
+      incrementNumMustCall(node);
+    }
+
+    if (skipTrackingInvocationResult(facts, node)) {
+      return;
+    }
+
+    if (typeFactory.hasMustCall(node.getTree())) {
+      incrementNumMustCall(node);
+    }
+    trackInvocationResult(facts, node);
+  }
+
+  /**
+   * If node is an invocation of a this or super constructor that has a MCA return type and an MCA
+   * parameter, check if any variable in facts is being passed to the other constructor. If so,
+   * remove it from facts.
```

Comment:
That's always true. I changed the comment when `facts` is first initialized in `analyze` to make this more clear. I'm not sure whether that's the most informative place to put it - it might be hard for someone reading just this method, for instance, to know to look there. Do you have any suggestions to help with that problem without needing to duplicate that comment on each method that modifies `facts`?

Reply:
Right, we definitely don't want it on every method.  (But every method that modifies `facts` should say that.)

There should be some overview documentation for the class as a whole that outlines the algorithm -- that could be a good place for this kind of information.

Reply:
Also, the name "facts" does imply "what is true".  Maybe a better name would be something more like "obligations" or a synonym for that word.

Reply:
I've made changes throughout to avoid the term `facts` and use `obligations` instead. I think the documentation is much better now; hopefully you will agree.